Commit ccr-stats telemetry device and tests

dliappis opened the pull request:
This PR adds a telemetry device called `ccr-stats`.

As this is meant to be used mostly internally no accompanying documentation is available yet, but I put some effort to document the Classes/Methods.

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+        """
+        :param telemetry_params: The configuration object for telemetry_params.
+            May optionally specify:
+            ``ccr-stats-indices``: JSON object specifying the indices per cluster to pushish statistics from.
```
This was meant to refer to https://www.w3schools.com/js/js_json_objects.asp and not the specific data type because we support reading both a JSON string directly in the cli arg or from a JSON filename.

danielmitterdorfer replied:
Ok

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+            Not all clusters need to be specified, but any name used must be be present in target.hosts.
+            Example:
+            {"ccr-stats-indices": {"cluster_a": ["follower"],"default": ["leader"]}
+            ``ccr-stats-sample-interval``: integer controlling the interval, in seconds, between collecting samples.
```
Sure will add the default value. Also +1 on positive (but what would a negative time interval mean?).
Btw I am already mentioning "between collecting samples", saying again "sampling interval" would be a repetition IMHO but I guess you suggested replacing the entire sentence.

danielmitterdorfer replied:
> I guess you suggested replacing the entire sentence.

yes, I did.

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+                        continue
+                    self.record_all_index_stats(index, stats[index])
+                except KeyError:
+                    logger.warning("The telemetry parameter 'ccr-stats-indices' specified the index [%s] for cluster [%s] but there are "
```
That's right, I am finding this useful because for CCR we aren't setting up the cluster with Rally and rely on custom runners to make an index start following an index in another already joined cluster.

danielmitterdorfer replied:
Ok.

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+
+    def record_all_index_stats(self, name, stats):
+        """
+        Push metric fields per shard to separate docs in metrics store.
```
Removed

dliappis commented on tests/mechanic/telemetry_test.py:
```diff
@@ -146,6 +148,14 @@ def info(self, *args, **kwargs):
 
 
+class TransportClient:
+    def __init__(self, stats=None):
```
Renamed

dliappis commented on tests/mechanic/telemetry_test.py:
```diff
@@ -200,6 +210,214 @@ def test_sets_options_for_java_9_or_above(self):
             env["ES_JAVA_OPTS"])
 
 
+class CCRStatsTests(TestCase):
```
Renamed

dliappis commented on tests/mechanic/telemetry_test.py:
```diff
@@ -200,6 +210,214 @@ def test_sets_options_for_java_9_or_above(self):
+        client = Client(transport_client=TransportClient(stats=ccr_stats_follower_response))
+        cfg = create_config()
+        metrics_store = metrics.EsMetricsStore(cfg)
+        recorder = telemetry.CCRStatsRecorder("remote", client, metrics_store, 1)
```
Done

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+            pass
+        time.sleep(self.sample_interval)
+
+    def record_all_index_stats(self, name, stats):
```
Renamed

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+        """
+
+        # ES returns all stats values in bytes or ms via "human: false"
+        stats = self.client.transport.perform_request('GET', '/_xpack/ccr/_stats', params={"human": "false"})
```
Fixed and also added a new test case.

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+    def __init__(self, cluster_name, client, metrics_store, sample_interval, indices=None):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The Elasticsearch client connection to the cluster.
```
Fixed

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+    help = "Regularly samples CCR related stats"
+
+    """
+    Gathers CCR stats on a cluster level
```
Done

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+            elif "_millis" in metric_name:
+                self.metrics_store.put_count_cluster_level(name=metric_name,
+                                                           count=metric_value,
+                                                           unit="millis",
```
Fixed

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+        :param stats: A dict with returned CCR stats for the index.
+        """
+
+        shard_metadata = {}
```
Done

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+                                   self.cluster_name)
+        else:
+            # Record metrics for every index returned by the CCR stats API.
+            if stats != {}:
```
Done

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+            # Record metrics only for indices specified with telemetry-params.
+            for index in self.indices:
+                try:
+                    if stats[index] == {}:
```
Fixed

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+                try:
+                    if stats[index] == {}:
+                        continue
+                    self.record_all_index_stats(index, stats[index])
```
Renamed

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+                if cluster_name not in clients.keys():
+                    raise exceptions.SystemSetupError(
+                        "The telemetry parameter 'ccr-stats-indices' must be a JSON Object with keys matching the cluster names "
+                        "specified in --target-hosts but it had [{}].".format(cluster_name))
```
Done (and adjusted the test case)

dliappis commented on esrally/mechanic/telemetry.py:
```diff
@@ -276,6 +276,159 @@ def detach_from_node(self, node, running):
+            Example:
+            {"ccr-stats-indices": {"cluster_a": ["follower"],"default": ["leader"]}
+            ``ccr-stats-sample-interval``: integer controlling the interval, in seconds, between collecting samples.
+        :param clients: A dict of client connections to all clusters.
```
Fixed

dliappis commented:
@danielmitterdorfer Thank you for the thorough review and your time! I addressed your comments and also added an additional test for https://github.com/elastic/rally/pull/505/files/040634296ce47e6e728e722657a271d6417de771#diff-c950a3c2ba2452023aed89b82cb2350c . I'd appreciate a second pass when you can.

danielmitterdorfer commented on esrally/mechanic/telemetry.py:
```diff
@@ -312,8 +312,8 @@ def __init__(self, telemetry_params, clients, metrics_store):
             for cluster_name in self.indices_per_cluster.keys():
                 if cluster_name not in clients.keys():
```
it should be sufficient to check with `if cluster_name not in clients`.

dliappis replied:
Fixed in https://github.com/elastic/rally/pull/505/commits/7be877663d6adde8d9a1e35bfb9eb83002a15274

danielmitterdorfer commented on tests/mechanic/telemetry_test.py:
```diff
@@ -234,11 +239,23 @@ def test_wrong_cluster_name_in_ccr_stats_indices_forbidden(self):
+
+
+class CcrStatsRecorderTests(TestCase):
+    @mock.patch("esrally.metrics.EsMetricsStore.put_count_cluster_level")
```
Do you need the mock here?

dliappis replied:
Yes we do, test fails without it. It's a good question though as I'd think execution would stop when we are raising `exceptions.RallyError` and the assertion consumes it (and stops there, thus not proceeding in `CcrStatsRecorder()` and reaching `put_value()`).

dliappis replied:
Scratch that. It works fine without mocking of course if I also remove the corresponding parameter in `test_raises_exception_on_transport_error()`

dliappis replied:
Addressed in https://github.com/elastic/rally/pull/505/commits/c9df447ec7fa9339d423ecf9ef6bf34261e15a82

danielmitterdorfer commented on tests/mechanic/telemetry_test.py:
```diff
@@ -234,11 +239,23 @@ def test_wrong_cluster_name_in_ccr_stats_indices_forbidden(self):
+
+class CcrStatsRecorderTests(TestCase):
+    @mock.patch("esrally.metrics.EsMetricsStore.put_count_cluster_level")
+    def test_raises_exception_on_transport_error(self, metrics_store_put_count):
```
Good test!

dliappis commented:
Many thanks for your review and perseverance @danielmitterdorfer !
